read_can_messages.py

```diff
@@ -0,0 +1,143 @@
+    datefmt="%Y-%m-%d %H:%M:%S",
+)
+
+
+def preprocess_data_format(format: Dict[str, List[Any]]) -> Dict[str, Dict[int, Any]]:
+    """
+    Parse data format and return in a more friendly format for  CAN consumption:
+    <CAN ID>: {
+        <offset>: {
+            <category 1>: ...
+            ...
+        }
+        ...
+    },
+    ...
+    """
+    processed = {}
+    for key, s in format.items():
+        # Get the arbitration ID and offset
+        can_id = int(s[-2], base=16)
+        offset = s[-1]
+        # Drop the ID and offset
+        s = s[:-2]
+        # Add/Update the message data in the processed data format
+        p = processed.setdefault(can_id, {})
+        num_bytes = s[0]
+        data_type = s[1]
+        units = s[2]
+        nominal_min = s[3]
+        nominal_max = s[4]
+        subsystem = s[5]
+        signal_info = SignalInfo(
+            key, num_bytes, data_type, units, nominal_min, nominal_max, subsystem
+        )
+        p[offset] = signal_info
+
+    return processed
+
+
+with open("sc1-data-format/format.json", "r") as file:
+    data = json.load(file)
+
+signal_definitions = preprocess_data_format(data)
+
+
+class MyListener(can.Listener):
+    def on_message_received(self, message):
+        message_data = {
+            "id": message.arbitration_id,
+            "data": list(message.data),
+            "timestamp": message.timestamp,
+        }
+
+        # loop through signal definitions to find can_id
+        can_id = message_data["id"]
+        if can_id not in signal_definitions:
+            logging.error(f"CAN ID {can_id:0x} not found in signal definitions.")
+            return
+        signals = signal_definitions[can_id]
+        byte_array = bytes(message.data)
+
+        for offset, signals_info in signals.items():
+            logging.debug(f"Processing signal at offset {offset} for CAN ID {can_id:0x}")
+            print(offset)
+            print(signals_info)
+            data_type = signals_info.type
+            signal_name = signals_info.name
+            if data_type == "float":
+                if len(byte_array) < 4:
+                    logging.error(
+                        f"Insufficient data for float signal in CAN ID {can_id:0x}."
+                    )
+                    return
+                else:
+                    # Unpack the first 4 bytes as a little-endian float.
+                    float_value = struct.unpack("<f", byte_array[:4])[0]
+                    logging.debug(
+                        f"New Message: ID={can_id:0x},Name={signal_name} Value={float_value}, Time Stamp={message_data['timestamp']}"
+                    )
+            elif data_type == "boolean":
+                bool_value = bool((byte_array[0] >> offset) & 1)
+                logging.debug(
+                    f"New Message: ID={can_id:0x},Name={signal_name} Value={bool_value}, Time Stamp={message_data['timestamp']}"
+                )
```

Comment:
Can we differentiate between all data types in the data format? Also, can we do a more strict check against the expected length in bytes for each value?